### PR TITLE
feat(DataLib): add option to allow non-matched subdata

### LINKED ISSUE #2491: Allow optional datalib's sub-data
Hello,

It should be great to allow optional sub-data if its corresponding mapped column/jsonpath is giving an error during the parse of the datalib's result set
This could be allowed by adding an "Optional" column in the sub-data table definition.

Regards
Aurélien

### PR DESCRIPTION
**Pull request by abourdon**
Please accept this PR that resolves #2491.
To avoid any breaking changes, I added an option to enable/disable allowing non-matched datalib's subdata and disabled this option by default.

Kind regards
Aurélien

### CODE REVIEW

File: source/src/main/java/org/cerberus/core/service/csvfile/ICsvFileService.java [hunk 1]
```diff
@@ -36,7 +36,18 @@ public interface ICsvFileService {
...
+	 *                                  Use {@link #IGNORE_NON_MATCHED_COLUMNS} &
+	 *                                  {@link #DO_NOT_IGNORE_NON_MATCHED_COLUMNS} for
+	 *                                  better readability
+	 * @param defaultNoMatchColumnValue the default value to set to any non-matched
+	 *                                  column if necessary
      * @return
      */
-    AnswerList<HashMap<String, String>> parseCSVFile(String urlToCSVFile, String separator, HashMap<String, String> columnsToGet,List<String> columnsToHide, TestCaseExecution execution);
+    AnswerList<HashMap<String, String>> parseCSVFile(String urlToCSVFile, String separator, HashMap<String, String> columnsToGet,List<String> columnsToHide, boolean ignoreNoMatchColumns, String defaultNoMatchColumnValue, TestCaseExecution execution);
+    boolean IGNORE_NON_MATCHED_COLUMNS = true;
```
Comment (vertigo17) [inline source/src/main/java/org/cerberus/core/service/csvfile/ICsvFileService.java:51]: Can we avoid creating those and use boolean values directly ?

File: source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java [hunk 2]
```diff
@@ -720,7 +728,7 @@ private AnswerList<HashMap<String, String>> getDataObjectList(TestDataLib lib, H
                 columnsToHide = getListOfSecrets(lib.getTestDataLibID());
 
                 // CSV Call is made here.
-                responseList = fileService.parseCSVFile(servicePathCsv, lib.getSeparator(), columnList, columnsToHide, execution);
+                responseList = fileService.parseCSVFile(servicePathCsv, lib.getSeparator(), columnList, columnsToHide, ignoreNonMatchedSubdata ? ICsvFileService.IGNORE_NON_MATCHED_COLUMNS : ICsvFileService.DO_NOT_IGNORE_NON_MATCHED_COLUMNS, defaultSubdataValue, execution);
```
Comment (vertigo17) [inline source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java:731]: That could be simplified sending just ignoreNonMatchedSubdata boolean.

File: source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java [hunk 3]
```diff
@@ -772,7 +780,7 @@ private AnswerList<HashMap<String, String>> getDataObjectList(TestDataLib lib, H
                                 Integer sqlTimeout = parameterService.getParameterIntegerByKey("cerberus_propertyexternalsql_timeout", system, 60);
 
                                 //performs a query that returns several rows containing n columns
-                                responseList = sqlService.queryDatabaseNColumns(connectionName, lib.getScript(), rowLimit, sqlTimeout, system, columnList, columnsToHide, execution);
+                                responseList = sqlService.queryDatabaseNColumns(connectionName, lib.getScript(), rowLimit, sqlTimeout, system, columnList, columnsToHide, ignoreNonMatchedSubdata ? ISQLService.IGNORE_NON_MATCHED_COLUMNS : ISQLService.DO_NOT_IGNORE_NON_MATCHED_COLUMNS, defaultSubdataValue, execution);
```
Comment (vertigo17) [inline source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java:783]: Same here.

File: source/src/main/java/org/cerberus/core/service/sql/ISQLService.java [hunk 4]
```diff
@@ -86,9 +86,20 @@ public interface ISQLService {
...
+	 *                                     for better readability
+	 * @param defaultNonMatchedColumnValue the default value to set to any
+	 *                                     non-matched column if necessary
      * @param execution
      * @return
      */
-    AnswerList<HashMap<String, String>> queryDatabaseNColumns(String connectionName, String sql, int rowLimit, int defaultTimeOut, String system, HashMap<String, String> columnsToGet, List<String> columnsToHide, TestCaseExecution execution);
-
+    AnswerList<HashMap<String, String>> queryDatabaseNColumns(String connectionName, String sql, int rowLimit, int defaultTimeOut, String system, HashMap<String, String> columnsToGet, List<String> columnsToHide, boolean ignoreNoMatchColumns, String defaultNonMatchedColumnValue, TestCaseExecution execution);
+    boolean IGNORE_NON_MATCHED_COLUMNS = true;
```
Comment (vertigo17) [inline source/src/main/java/org/cerberus/core/service/sql/ISQLService.java:103]: Here as well.

File: source/src/main/java/org/cerberus/core/service/csvfile/ICsvFileService.java (hunk 1 above)
Comment (abourdon) [reply source/src/main/java/org/cerberus/core/service/csvfile/ICsvFileService.java]: These variables were used for better readability. The latest commit (#17ba4e0) removes them.

File: source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java (hunk 2 above)
Comment (abourdon) [reply source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java]: These variables were used for better readability. The latest commit (#17ba4e0) removes them.

File: source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java (hunk 3 above)
Comment (abourdon) [reply source/src/main/java/org/cerberus/core/service/datalib/impl/DataLibService.java]: These variables were used for better readability. The latest commit (#17ba4e0) removes them.

File: source/src/main/java/org/cerberus/core/service/sql/ISQLService.java (hunk 4 above)
Comment (abourdon) [reply source/src/main/java/org/cerberus/core/service/sql/ISQLService.java]: These variables were used for better readability. The latest commit (#17ba4e0) removes them.

**Pull request merged**